### ISSUE TITLE
Research: Bring interdependencies of `systemd` services to containerized AGW

### ISSUE THREAD

**Issue by mpfirrmann** [magma/magma#14055]
The different magma services are interconnected, i.e. if one service restarts, some others do as well (see also this [cool diagram](https://excalidraw.com/#room=9b3ea34b6f337ccab568,F2DMzJxlW8yjKZtjQ2djtA)). In `systemd` this is solved in the config-`.yaml` files and restarting one service i.e. during the S1AP tests (handled by `s1ap_utils.py`) just does the job. This is not the case with a containerized AGW. Here, the dependencies are only hard coded in the restart function, which leads to ugly if-else constructions. While this works in a test environment, these dependencies are not represented in production.
Follow up to #13965 and #13966.

Open questions:
1. How exactly are the dependencies between the services (see the linked diagram)? Which services need to be restarted for things to work properly, and which services must *not* be restarted? (Would it be an option to just always restart all services?)
  - [x] Document results
    - Check the documentation if there is something about the reason of these restarts
    - Look at integration tests to understand the reason for the restarts, e.g. by removing them and breaking tests
2. How do you realize these dependencies in a docker-compose environment? Move the logic inside magmad?
  - [ ] Contact people responsible for dockerization, find out if a solution for this problem was already designed
  - [x] Document possible solutions in docker-compose, kubernetes, magmad

**Comment by ajahl** [comment]
| Service           | Before           | After 		    | Requires 	      | Wants 		| PartOf           | 
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
| MME               |  			   | MobilityD	     |  SctpD           |                        |  SessionD      |
|                        | 			   | PipelineD	     |  	              |                        |  MobilityD	 |
|                        | 			   | SessionD	     |  	              |                        |  PipelineD	 |
|                        | 			   | SctpD		     |  	              |                        |  			 |
|                        | 			   |    		     |  	              |                        |  			 |
| ConnectionD |  MME		   | PipelineD	     |                        |                        |  MME             | 
|                        | 			   | OVS   		     |  	              |                        |  			 |
|                        | 			   |    		     |  	              |                        |  			 |
| SessionD       |  MME    	   |    		     |  	              |                        |  MME		 |
|                        | 			   |    		     |  	              |                        |  			 |
| MobilityD       |  MME		   | OVS    	     |                        | OVS                |  MME             | 
|                        | 			   | SubscriberDB |  	              | SubscriberDB |  			 |
|                        | 			   |    		     |  	              |                        |  			 |
| PipelineD       |  MME		   | OVS    	     |                        | OVS                |  MME             | 
|                        | 			   | MobilityD        |  	              |                        |  MobilityD	 |
|                        | 			   |    		     |  	              |                        |  OVS		 |
|                        | 			   |    		     |  	              |                        |  			 |
| liAgent           | 		    	   |    PipelineD     |  	              |                        |  PipelineD	 |
|                        | 			   |    		     |  	              |                        |  			 |
| SctpD            | 	MME   	   |                  	     |  	      	      |                        | 	                 |
|                        | 			   |    		     |  	              |                        |  			 |

**Comment by ajahl** [comment]
Possible solution approaches to realise service dependencies in `docker-compose`, `Kubernetes`, and `Magmad`.

- In `Docker-compose`, dependencies regarding the services starting order can be defined using `depends_on`. However, this only works when starting and stopping services with `docker-compose`, not when restarting individual services.

- `Kubernetes` made no distinction between containers in a pod by default. It is not possible to control which container starts first or terminates last. Nevertheless, Kubernetes provides two special container types. Both types must run in the same pod as the containers that depend on. 
  -  `InitContainers` define containers that are started before the normal containers. They are started sequentially and must run completely before the pod is ready to start the normal containers. These containers only influence the initial start-up process. Their resource requirements and limitations are handled differently.
  - `Sidecar` containers also start before the normal containers but are shut down after all other containers have finished. They behave like normal containers. The only difference is their life cycle.

  Dependencies across multiple services following a dependency graph cannot be mapped with both container types.

- In `Magmad`, a solution would be a watchdog that monitors `docker-compose events` (active polling) and restarts corresponding services on detected restart events.

**Comment by ajahl** [comment]
Except for Sctpd, restarting dependent services does not seem to be necessary in containerised AGW. On my local machine, running integ tests produces the same results with and without additional service restarts. Also executing [integ tests in the CI](https://github.com/ajahl/magma/actions/runs/3256317211/jobs/5346616243) shows the same output.

**Comment by Neudrino** [comment]
The relation of the **before** and **after** column from above table.

```mermaid
graph TD

    MobilityD --> MME
    PipelineD --> MME
    SessionD --> MME
    SctpD --> MME

    PipelineD --> ConnectionD
    OVS --> ConnectionD
    ConnectionD --> MME

    OVS --> MobilityD
    SubscriberDB --> MobilityD

    OVS --> PipelineD
    MobilityD --> PipelineD

    PipelineD --> LiAgent

    EnvoyController --> Envoy
    EnvoyController --> PipelineD
```

**Comment by Neudrino** [comment]
```mermaid
graph TD
    MME ==> MobilityD
    MME ==> PipelineD
    MME ==> SessionD
    MME --> SctpD

    ConnectionD ==> MME
    SessionD ==> MME 

    MobilityD -.-> OVS 
    MobilityD -.-> SubscriberDB
    MobilityD ==> MME 

    PipelineD -.-> OVS
    PipelineD ==> MME
    PipelineD ==> MobilityD
    PipelineD ==> OVS

    LiAgent ==> PipelineD

    Envoy ==> PipelineD
    EnvoryController --> Envoy
    EnvoyController ==> PipelineD
```

**Comment by Neudrino** [comment]
### Combined graph for `PartOf` relations only.

```mermaid
graph TD
    MME ==> MobilityD
    MME ==> PipelineD
    MME ==> SessionD

    MobilityD ==> MME 
    ConnectionD ==> MME
    SessionD ==> MME 

    PipelineD ==> MME
    PipelineD ==> MobilityD
    PipelineD ==> OVS

    LiAgent ==> PipelineD

    Envoy ==> PipelineD
    EnvoyController ==> PipelineD
```

**Comment by Neudrino** [comment]
### Combined graph for `Wants` & `Requires` relations only.

```mermaid
graph TD
    MME --> SctpD

    MobilityD -.-> OVS 
    MobilityD -.-> SubscriberDB

    PipelineD -.-> OVS
    EnvoryController --> Envoy
```

**Comment by voisey** [comment]
@wolfseb Assigning you to see whether you can reproduce the behaviour you said you observed in the past regarding interdependent container restarts.

**Comment by wolfseb** [comment]
In my testing I found the following behavior:
* Run any test (even the basic `test_attach_detach` is enough)
* After the test finishes, restart mme with `docker-compose restart oai_mme`
* mme is now in a restart loop and unable to come up properly
* Restarting mobilityd with `docker-compose restart mobilityd` lets mme come up again on its next restart
* After a restart of mme and mobilityd, mme can be restarted alone without any issue. It seems like there is some data left in mobilityd after a test run that causes issues. Restarting only mme within the tests works fine though (I just tried this with `test_attach_ul_udp_data_with_mme_restart`).
